From d2f038cdba2d47e20c771fb5c5c98a8fd4c9bccd Mon Sep 17 00:00:00 2001
From: Sibren Vasse <github@sibrenvasse.nl>
Date: Mon, 3 Jun 2019 20:02:44 +0200
Subject: [PATCH] Revert "Validate required parameters in
 inet6_validate_link_af"

This reverts commit 7dc2bccab0ee37ac28096b8fcdc390a679a15841.
---
 net/ipv6/addrconf.c | 57 +++++++++++++++++----------------------------
 1 file changed, 22 insertions(+), 35 deletions(-)

diff --git a/net/ipv6/addrconf.c b/net/ipv6/addrconf.c
index 081bb517e40d..e2d9fa0c98cb 100644
--- a/net/ipv6/addrconf.c
+++ b/net/ipv6/addrconf.c
@@ -5657,6 +5657,18 @@ static const struct nla_policy inet6_af_policy[IFLA_INET6_MAX + 1] = {
 	[IFLA_INET6_TOKEN]		= { .len = sizeof(struct in6_addr) },
 };
 
+static int inet6_validate_link_af(const struct net_device *dev,
+				  const struct nlattr *nla)
+{
+	struct nlattr *tb[IFLA_INET6_MAX + 1];
+
+	if (dev && !__in6_dev_get(dev))
+		return -EAFNOSUPPORT;
+
+	return nla_parse_nested_deprecated(tb, IFLA_INET6_MAX, nla,
+					   inet6_af_policy, NULL);
+}
+
 static int check_addr_gen_mode(int mode)
 {
 	if (mode != IN6_ADDR_GEN_MODE_EUI64 &&
@@ -5677,44 +5689,14 @@ static int check_stable_privacy(struct inet6_dev *idev, struct net *net,
 	return 1;
 }
 
-static int inet6_validate_link_af(const struct net_device *dev,
-				  const struct nlattr *nla)
-{
-	struct nlattr *tb[IFLA_INET6_MAX + 1];
-	struct inet6_dev *idev = NULL;
-	int err;
-
-	if (dev) {
-		idev = __in6_dev_get(dev);
-		if (!idev)
-			return -EAFNOSUPPORT;
-	}
-
-	err = nla_parse_nested_deprecated(tb, IFLA_INET6_MAX, nla,
-					  inet6_af_policy, NULL);
-	if (err)
-		return err;
-
-	if (!tb[IFLA_INET6_TOKEN] && !tb[IFLA_INET6_ADDR_GEN_MODE])
-		return -EINVAL;
-
-	if (tb[IFLA_INET6_ADDR_GEN_MODE]) {
-		u8 mode = nla_get_u8(tb[IFLA_INET6_ADDR_GEN_MODE]);
-
-		if (check_addr_gen_mode(mode) < 0)
-			return -EINVAL;
-		if (dev && check_stable_privacy(idev, dev_net(dev), mode) < 0)
-			return -EINVAL;
-	}
-
-	return 0;
-}
-
 static int inet6_set_link_af(struct net_device *dev, const struct nlattr *nla)
 {
+	int err = -EINVAL;
 	struct inet6_dev *idev = __in6_dev_get(dev);
 	struct nlattr *tb[IFLA_INET6_MAX + 1];
-	int err;
+
+	if (!idev)
+		return -EAFNOSUPPORT;
 
 	if (nla_parse_nested_deprecated(tb, IFLA_INET6_MAX, nla, NULL, NULL) < 0)
 		BUG();
@@ -5728,10 +5710,15 @@ static int inet6_set_link_af(struct net_device *dev, const struct nlattr *nla)
 	if (tb[IFLA_INET6_ADDR_GEN_MODE]) {
 		u8 mode = nla_get_u8(tb[IFLA_INET6_ADDR_GEN_MODE]);
 
+		if (check_addr_gen_mode(mode) < 0 ||
+		    check_stable_privacy(idev, dev_net(dev), mode) < 0)
+			return -EINVAL;
+
 		idev->cnf.addr_gen_mode = mode;
+		err = 0;
 	}
 
-	return 0;
+	return err;
 }
 
 static int inet6_fill_ifinfo(struct sk_buff *skb, struct inet6_dev *idev,
-- 
2.21.0

